### PR TITLE
Rails upgrade to 5.2

### PR DESCRIPTION
**Pull request by pjaspers**
- [x] Upgrade to 5.1
- [x] Upgrade to 5.2

## Open questions

`gem "easy_diff" # No longer in use?` Is this gem still in use or can we remove it?

## Some notes

- FactoryGirl -> FactoryBot
- Papertrail to latest version
- Sentry to latest version
- Sidekiq went from 4 to 5
- Devise from 4.4 to 4.5
- Gemfile reordered into sections
- More gems being locked to a version
- Seeding fixed (again) 

## Self proof reading checklist

- [x] Did I run Pronto and fixed all warnings
- [x] Update documentation,readme accordingly

### CODE REVIEW

**Comment by mestachs** [comment]
is this still a WIP or should I give it a try ?

**Comment by pjaspers** [comment]
There are some development gems that could use an update (e.g. annotate), I would like to move Rails Admin to a later version (but that will pull in a new version of jquery-ui and I need to figure out if we use that?)

But most of the update is done, I'm deploying a review app now, and will ping you when done.

**Comment by pjaspers** [comment]
@mestachs I consider this PR up to date, it takes the current application and upgrades Rails (and a lot fo the gem dependencies) to their latest version.

It's currently deployed at https://orbf2-staging-pr-89.herokuapp.com with the seeds already run.

File: app/models/payment_rule.rb [hunk 1]
```diff
@@ -19,7 +27,7 @@ class PaymentRule < ApplicationRecord
   has_one :rule, dependent: :destroy, inverse_of: :payment_rule
   has_many :package_payment_rules, dependent: :destroy
   has_many :packages, through: :package_payment_rules, source: :package
-  has_many :datasets, dependent: :destroy, class_name: PaymentRuleDataset
+  has_many :datasets, dependent: :destroy, class_name: "PaymentRuleDataset"
```
Comment (mestachs) [inline app/models/payment_rule.rb]: codeclimate see to we probably need to "Specify an `:inverse_of` option."

File: config/environments/production.rb [hunk 2]
```diff
@@ -34,6 +38,12 @@
   # config.action_dispatch.x_sendfile_header = 'X-Sendfile' # for Apache
   # config.action_dispatch.x_sendfile_header = 'X-Accel-Redirect' # for NGINX
 
+
+  if ENV["ENABLE_ACTIVE_STORAGE"]
+    # Store uploaded files on the local file system (see config/storage.yml for options)
+    config.active_storage.service = :local
```
Comment (mestachs) [inline config/environments/production.rb]: that looks dangerous to me to keep that in "production.rb", will silently upload and files will vanish on recycling/deployment

File: config/puma.rb [hunk 3]
```diff
@@ -1,13 +1,13 @@
 # Puma can serve each request in a thread from an internal thread pool.
-# The `threads` method setting takes two numbers a minimum and maximum.
+# The `threads` method setting takes two numbers: a minimum and maximum.
 # Any libraries that use thread pools should be configured to match
 # the maximum value specified for Puma. Default is set to 5 threads for minimum
-# and maximum, this matches the default thread size of Active Record.
+# and maximum; this matches the default thread size of Active Record.
 #
-threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }.to_i
+threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }
```
Comment (mestachs) [inline config/puma.rb]: don't think we have RAILS_MAX_THREADS specified but if it is, thread_count will be a string no ?

ENV.fetch("RAILS_MAX_THREADS", "5" ).to_i ?

File: Gemfile [hunk 4]
```diff
@@ -7,55 +9,75 @@ git_source(:github) do |repo_name|
...
+gem "rails-jquery-autocomplete"
+gem "sassc-rails", "~> 2.0.0"
+gem "simple_form"
+gem "sprockets", "~> 3.7.2"
+gem "turbolinks", "~> 5"
+gem "uglifier", ">= 1.3.0"
+
+## Authentication
+
+gem "devise", "~> 4.5.0"
```
Comment (mestachs) [inline Gemfile]: did you test the signin/signon/forget screens ?

File: lib/tasks/auto_annotate_models.rake [hunk 5]
```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+# NOTE: only doing this in development as some production environments (Heroku)
+# NOTE: are sensitive to local FS writes, and besides -- it's just not proper
+# NOTE: to have a dev-mode tool do its thing in production.
+if Rails.env.development?
+  require "annotate"
```
Comment (mestachs) [inline lib/tasks/auto_annotate_models.rake]: do we still launch with annotate command ? or is there a specify task to run ?

File: config/environments/production.rb (hunk 2 above)
Comment (pjaspers) [reply config/environments/production.rb]: I've been on the fence about actually keeping the whole `ENV["ENABLE_ACTIVE_STORAGE"]` thing. I'm just going to remove it and let a future PR readd it.

File: config/puma.rb (hunk 3 above)
Comment (pjaspers) [reply config/puma.rb]: Yes, thought the same thing, but the rails generator overrode this. I will re-overwrite it.

File: Gemfile (hunk 4 above)
Comment (pjaspers) [reply Gemfile]: I did not.

File: lib/tasks/auto_annotate_models.rake (hunk 5 above)
Comment (pjaspers) [reply lib/tasks/auto_annotate_models.rake]: From now on they run automatically when you run your migrations, so normally everything is kept up to date.

If you want to run them manually `bundle exec annotate` will still work as expected.

File: config/puma.rb (hunk 3 above)
Comment (mestachs) [reply config/puma.rb]: may be puma change the way they use it : actual_thread_count = Integer(threads_count)

File: lib/tasks/auto_annotate_models.rake (hunk 5 above)
Comment (mestachs) [reply lib/tasks/auto_annotate_models.rake]: That's even better... thanks for the setup. Can I propose you explain this in next dev meeting so we apply this on other projects too.
Wondering if we can automate running some tools like the one here : https://github.com/mestachs/experiment/blob/master/codereview/migration_review_checklist.md

Comment (pjaspers) [reply lib/tasks/auto_annotate_models.rake]: I've used that `strong_migrations` gem in the past, but it got removed later on, I don't remember why exactly. I think it was  too much maintenance when writing a migration.

I also used [danger](https://github.com/danger/danger) once, which we could use to help adding (automated) feedback on PRs.

File: Gemfile (hunk 4 above)
Comment (pjaspers) [reply Gemfile]: Now checked, they seem to check out.

**Pull request merged**